docs/device-assignment.md

```diff
@@ -0,0 +1,130 @@
+
+The main difficulty of this approach is reliably identifying which system device was assigned. The KubeVirt API must be verbose enough to let users request a device by it's vendor ID/product ID tuple, yet hide the actual device address. This is difficult for libvirt, as VM XML requires users to specify the device address.
+
+Because the actual device selection is done by kube-scheduler, and system may contain multiple devices with identical vendor ID/product ID that are in the same or different IOMMU groups, we need a mechanism to inspect which IOMMU group does the device belong to.
+
+To demonstrate the issue, we present an example.
+
+### Example
+
+* IOMMU group `1` contains devices
+ *	`A` (vendor ID = 1000, product ID = 1000)
+ * `B` (vendor ID = 2000, product ID = 2000)
+* IOMMU group `2` contains
+ * device `C` (vendor ID = 1000, product ID = 1000)
+* we have VM `α`, and it's corresponding pod `a`
+
+```yaml
+apiVersion: kubevirt.io/v1alpha1
+kind: VirtualMachine
+metadata:
+  name: α
+spec:
+  domain:
+    devices:
+      ...
+      passthrough:
+      - type: pci
+        vendor: 1000
+        device: 1000
+      ...
+    memory:
+      unit: MB
+      value: 64
+    os:
+      type:
+        os: hvm
+    type: qemu
+```
+
+In this case, KubeVirt must generate pod specification that requests device with vendor ID/product ID 1000:1000. Such specification is shown below.
+
+```yaml
+spec:
+  containers:
+  - name: a
+    ...
+    resources:
+      requests:
+              mpolednik.github.io/1000_1000: 1
+      limits:
+              mpolednik.github.io/1000_1000: 1
+```
+
+The difficult part is: how should libvirt XML generated by virt-handler look like? It is able to query the underlying system for devices and determine addresses of all devices that match vendor ID/product ID 1000:1000, but choosing a different device than the scheduler scheduled would cause the VM not to start due to IOMMU violation.
+
+To safely identify the selected device, we have to extend `virt-launcher` to report back the IOMMU groups inside the VMs pod. This could be done via a file in a directory shared between virt-launcher and virt-handler. Knowing which IOMMU groups were assigned to the pod, virt-handler can then probe the system and find device addresses to construct the final libvirt XML.
+
+In case of selecting N similar devices from IOMMU group containing M similar devices, where N <= M, the device ordering is not relevant.
+
+### API
+
+The API must be able to abstract devices to the level consumable by cloud: an identifier for device that contains enough information that the user can understand what the device is. The tuple vendor ID/product ID does exactly that - all of the device addressing is abstracted while the information can be used to query the exact model of the device (e.g. from PCI IDs).
+
+```yaml
+apiVersion: kubevirt.io/v1alpha1
+kind: VirtualMachine
+metadata:
+  name: α
+spec:
+  domain:
+    devices:
+      ...
+      passthrough:
```

Comment:
You might want to pick up the "struct  per type" API approach with the reasoning that there are different properties depending on the passthrough device type, i.e.:
```
passthrough:
- pci:
    vendor: ..
    device: ..
    managed: true
- scsi:
    vendor: ..
    device: ..
    sgio: filtered
```

The next change would be to not spell out the vendor+device, but rather reference the resource class directly:

```
passthrough:
- pci:
    resource: mpolednik.github.io/1000_1000
    managed: true
- scsi:
    resource: mpolednik.github.io/1000_1042
    sgio: filtered
```

Reason for this is to make a transparent association between the VM and resource class.

Reply:
For the first idea, we have to be careful which libvirt knobs do we expose. It does not make sense to let the user set e.g. `managed` as that is unusable with kubernetes DPIs. Referencing resource classes directly sounds fine, although it's quite close because the vendor/device will have to be parsed out of the name. For the sake of consistence with storage, it's probably better - will incorporate that into next revision.